src/lib.rs

```diff
@@ -39,4 +62,76 @@ mod tests {
         // THEN it should be zero
         assert_eq!(size, 0);
     }
-}
+
+    #[test]
+    fn it_should_report_the_correct_number_of_elements_when_filled() {
+        //Given an empty buffer
+        let mut buffer: RingBuffer<i32> = RingBuffer::<i32>::new(10);
+        //when added 4 elements
+        for i in 0..4 {
+            buffer.add(i);    
+        }
+        let size = buffer.number_of_elements();
+        //then it should return 4
+        assert_eq!(size, 4);
+    }
+
+    #[test]
+    fn it_should_report_the_correct_number_of_elements_when_overfilled() {
+        //Given an empty buffer
+        let mut buffer: RingBuffer<i32> = RingBuffer::<i32>::new(10);
+        //when added 12 elements
+        for i in 1..13 {
+            buffer.add(i);    
+        }
+        let size = buffer.number_of_elements();
+        //then it should return 10
+        assert_eq!(size, 10);
+    }
+
+    #[test]
+    fn it_should_return_the_before_pushed_values() {
+        //Given an empty buffer
+        let mut buffer: RingBuffer<i32> = RingBuffer::<i32>::new(10);
+        //when added 12 elements
+        for i in 1..11 {
+            println!("added {i}");
+            buffer.add(i);    
+        }
+        //then it should return 10
+        for i in 1..11 {
+            let ret_val = buffer.remove();
+            println!("expected {}, was {:?}", i , ret_val);
+            assert_eq!(ret_val, Some(i)); //Why has this to be some of...? 
+        }
+    }
+
+    #[test]
+    fn it_should_return_none_when_empty() {
```

Comment:
very nice that you thought about this situation, and that you even added a test for it. the test looks trivial but it serves so many purposes.

I did not need to look at your code to understand the bahvior of the ringbuffer, your tests gave me plenty of examples to understand it's behavior! :+1: :+1: :+1: